--- a/net/minecraft/client/renderer/LightTexture.java
+++ b/net/minecraft/client/renderer/LightTexture.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.level.dimension.DimensionType;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
+import net.neoforged.neoforge.client.renderer.IGraphicsShader;
 import org.joml.Vector3f;
 
 @OnlyIn(Dist.CLIENT)
@@ -103,7 +_,7 @@
                 float f8 = clientlevel.dimensionType().ambientLight();
                 boolean flag = clientlevel.effects().forceBrightLightmap();
                 float f9 = this.minecraft.options.gamma().get().floatValue();
-                CompiledShaderProgram compiledshaderprogram = Objects.requireNonNull(RenderSystem.setShader(CoreShaders.LIGHTMAP), "Lightmap shader not loaded");
+                IGraphicsShader compiledshaderprogram = Objects.requireNonNull(RenderSystem.setShader(CoreShaders.LIGHTMAP), "Lightmap shader not loaded");
                 compiledshaderprogram.safeGetUniform("AmbientLightFactor").set(f8);
                 compiledshaderprogram.safeGetUniform("SkyFactor").set(f1);
                 compiledshaderprogram.safeGetUniform("BlockFactor").set(f7);
@@ -141,7 +_,7 @@
     }
 
     public static int block(int p_109884_) {
-        return p_109884_ >>> 4 & 15;
+        return (p_109884_ & 0xFFFF) >> 4; // Forge: Fix fullbright quads showing dark artifacts. Reported as MC-169806
     }
 
     public static int sky(int p_109895_) {
